### ISSUE TITLE
Beckhoff: correctly work out if axis is moving

### ISSUE THREAD

**Issue by rerpha** [ISISComputingGroup/IBEX#6874]
As a scientist/developer I would like to be correctly told if a motor is moving. Currently we use the BMOVING flag which isn't appropriate as it doesn't take into account post move corrections etc. 

At the time of writing this ticket we're unsure what to use instead, @Simon-Cooper is talking to Beckhoff. 

### Acceptance Criteria
- [ ] tcioc's motor record is modified to point at this flag instead
- [ ] tests are updated

### Extra Information
Other things such as whether all motors are moving etc should still work because they're plumbed into the motor record's MOVN flag.

**Comment by DominicOram** [comment]
Additionally we found that with a simulated PLC on move start the BMOVING takes a while to update, this causes MOVN to flip flop between 0 and 1 for a while.

**Comment by Simon-Cooper** [comment]
For reference, my understanding from discussion with Beckhoff is that the bMoving signal is only ON when the setpoint position of the NC axis is changing (nothing to do with the actual motor position). I guess it's assumed that your actual position will match the setpoint trajectory well enough that this isn't a problem but that might not be the case for us.

**Comment by rerpha** [comment]
I have successfully managed to set up a scope and i can see that while the fActPosition is changing bMoving is pretty much set to 1 and never oscillates so I think the problem is definitely something in tcioc. I am going to try with a single axis and see if it consistently takes a few seconds to set the moving flag

**Comment by rerpha** [comment]
```
	// Check if it's time to do an EPICS read for the slow (read only) records
	bool readAll = false;
	if (cyclesLeft <= 0) readAll = true;
	// Reset countdown until EPICS read
	if (readAll) cyclesLeft = scanRateMultiple;
```

JUst spotted this in the code - bmoving would be read only i guess..

**Comment by rerpha** [comment]
ok, the above never actually gets called, for whatever reason it thinks that the tc:// calls are read + write only. I guess that's fine and probably one less thing to go wrong as they'll all be called at the faster rate

**Comment by rerpha** [comment]
Next I am going to have a play with https://github.com/stlehmann/pyads and monitor the `bmoving` flag when a setpoint is set through EPICS to rule out that it's 100% not an ADS issue and rather a tcioc internal one

**Comment by rerpha** [comment]
From what i've noticed it takes pretty much dead on 5.5 seconds for the moving flag to update...

**Comment by rerpha** [comment]
Another update, it's not just the moving flag that is not updated it's ANYTHING that isn't updated with  a different value very quickly - FVELOCITY (LREAL type not BOOL) also does not get updated quickly on one set. 

So I wrote a test to see if updating it really quickly does actually process the PV update: 

```python 
import pyads
import time
from datetime import datetime
SLEEP_TIME = 1
# connect to plc and open connection
plc = pyads.Connection('192.168.0.25.1.1', 852)
with plc:
    while True:
        for i in range(1, 10):
            # read int value by name
            plc.write_by_name("MAIN.afbAxes[1].stAxisStruct.stControl.fVelocity", i)
            print(f"Sent {i} velocity to PLC at {datetime.now()}")
            time.sleep(SLEEP_TIME)

```


this outputs as follows: 

```
Sent 1 velocity to PLC at 2021-11-26 10:18:10.515402
Sent 2 velocity to PLC at 2021-11-26 10:18:11.518399
Sent 3 velocity to PLC at 2021-11-26 10:18:12.521399
Sent 4 velocity to PLC at 2021-11-26 10:18:13.524410
Sent 5 velocity to PLC at 2021-11-26 10:18:14.528439
Sent 6 velocity to PLC at 2021-11-26 10:18:15.532410
Sent 7 velocity to PLC at 2021-11-26 10:18:16.536412
Sent 8 velocity to PLC at 2021-11-26 10:18:17.540411
Sent 9 velocity to PLC at 2021-11-26 10:18:18.543399
Sent 1 velocity to PLC at 2021-11-26 10:18:19.546415
Sent 2 velocity to PLC at 2021-11-26 10:18:20.550396
Sent 3 velocity to PLC at 2021-11-26 10:18:21.554391

```

which almost matches a `camonitor` of the `FVELOCITY` PV; 

```
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:17:59.946435 4
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:11.579420 1
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:12.588402 2
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:13.546412 3
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:14.555438 4
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:15.565396 5
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:16.576397 6
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:17.588397 7
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:17.790413 8
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:19.611416 9
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:20.573398 1
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:21.585407 2
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:18:26.738374 3
```

as you can see they are pretty much updated every second. 



now change SLEEP_TIME to 5 (send velocity every 5 seconds) and you get this: 


```
Sent 1 velocity to PLC at 2021-11-26 10:20:14.190385
Sent 2 velocity to PLC at 2021-11-26 10:20:19.194383
Sent 3 velocity to PLC at 2021-11-26 10:20:24.197383
Sent 4 velocity to PLC at 2021-11-26 10:20:29.200384
Sent 5 velocity to PLC at 2021-11-26 10:20:34.204367
Sent 6 velocity to PLC at 2021-11-26 10:20:39.207366
```

and then inconsistent PV updates: 


```
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:19:55.940227 3
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:19.250384 1
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:22.684370 2
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:29.258369 3
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:31.586381 4
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:39.273381 5
TE:NDLT1542:ASTAXES_1:STCONTROL-FVELOCITY.RVAL 2021-11-26 10:20:40.535381 6
```

**Comment by rerpha** [comment]
setting the dirty flag on all PLC records in tcComms.cpp rather than a small subset of them fixes the delay from ASD, which I think means we could just modify that or add a flag to make it configurable - however the motor record still tends to do something weird when receiving setpoints so i'm going to look into that next. I think this is because when it receives a setpoint it sets MOVN to 1 then bmoving hasn't updated yet which forces it back to 0, then 1 again. This only happens sometimes, as it's a bit of a race condition between tcioc and the motor record. We could either:
A) live with this, which measn that waitfor_moves etc won't work nicely 
B) change the implementation of the motor record to look at bmoving entirely if possible for deciding when to set MOVN.

I think the latter is the better approach so i'm going to see if that's possible.

**Comment by rerpha** [comment]
message to myself - this is the fix/workaround for the delay: https://github.com/ISISComputingGroup/EPICS-tcIoc/pull/12

to do: sort out the weird flip flopping now for DMOV and MOVN - there is now no (or a very short) delay for moving, just need to work out how to remedy the flip flopping between 0-1-0-1 on initial move.

**Comment by rerpha** [comment]
repointing to 8 for next sprint, think it'll take a couple of days to sort out the MOVN flipflopping as there's something weird going on with timestamping.  probably another day for review

**Comment by FreddieAkeroyd** [comment]
The above mentions both `MOVN` and `DMOV`. Can you confirm that `MOVN` is a nicety and not a problem, and it is just `DMOV` that needs to be fixed? I believe `waitfor_move` etc just use `DMOV`

**Comment by rerpha** [comment]
> The above mentions both `MOVN` and `DMOV`. Can you confirm that `MOVN` is a nicety and not a problem, and it is just `DMOV` that needs to be fixed? I believe `waitfor_move` etc just use `DMOV`

fairly sure DMOV was affected but will have to double check, will get it set up now

**Comment by FreddieAkeroyd** [comment]
I think it is up to the motor driver, some may write the same value to both DMOV and MOVN, but we shouldn't be relying on MOVN anywhere.

**Comment by rerpha** [comment]
Have just tried it, there is a slight ordering issue (DMOV updating before BMOVING, but only ~60ms), but dmov seems to not be affected. weird, thought it was, but maybe I was looking at MOVN mistakenly. In which case I think this ticket may be done.. 

I will continue debugging for a bit with the maintainer but for now we at least have a workaround and i'd like to get it in for this release so going to tidy this up and move it into review

**Comment by rerpha** [comment]
PRs: 

tcioc: https://github.com/ISISComputingGroup/EPICS-tcIoc/pull/12
release notes: https://github.com/ISISComputingGroup/IBEX/pull/6952/files

To test: 

set up a simulated Beckhoff PLC using everything on master, start TWINCAT IOC and monitor these PVs: 
```
camonitor %MYPVPREFIX%MOT:MTR0101.DMOV %MYPVPREFIX%MOT:MTR0101 %MYPVPREFIX%ASTAXES_1:STSTATUS-BMOVING.RVAL %MYPVPREFIX%MOT:MTR0101.RBV
```
perform a move either through the table of motors or through the Beckhoff main VISU in visual studio ( PLC -> tc_project_app -> tc_project_app Project -> tc_mca_std_lib -> VISUs -> MainVisu )

you should see that there is a huge delay on the BMOVING flag compared to an actual move

now check out tcioc branch and `make` from the top level

do the same again and you should see the moving updates almost immediately. DMOV should also flip to 0 when moving then 1 when done.

**Comment by FreddieAkeroyd** [comment]
Does the motor record poll loop use the same axis status read as the twincat update of the BMOVING PV? If so, it may be just the motor record poll that updates the PV is running more often.

**Comment by rerpha** [comment]
yes, I think so, its done in pollAll 
could also be to do with hardware timestamps vs epics motor record timestamps? either way im not sure its that important

**Comment by FreddieAkeroyd** [comment]
you could run `camonitor -t cs  PV_OF_INTEREST` - this prints the client as well as server side timestamps. It may be both pvs update at the same time, but the BMOVING gets a hardware timestamp and DMOV get the server software timestamp. If you add `-cs` to camonitor you'll get the local timestamp when the update happened, which might give more of a clue as if the updates get posted at the same time

**Comment by rerpha** [comment]
here is the output of using -t cs

```
TE:NDLT1542:MOT:MTR0101.DMOV   2022-01-04 14:42:00.705298(2022-01-04 14:42:00.706364) 0
TE:NDLT1542:ASTAXES_1:STSTATUS-BMOVING.RVAL 2022-01-04 14:42:00.783923(2022-01-04 14:42:00.783106) 1
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:00.905118(2022-01-04 14:42:00.905152) 151.193
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:01.106074(2022-01-04 14:42:01.106166) 154.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:01.306290(2022-01-04 14:42:01.306115) 155.793
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:01.507423(2022-01-04 14:42:01.507111) 157.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:01.707637(2022-01-04 14:42:01.707088) 160.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:01.907912(2022-01-04 14:42:01.907148) 161.793
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:02.108106(2022-01-04 14:42:02.107281) 163.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:02.309237(2022-01-04 14:42:02.308426) 166.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:02.509420(2022-01-04 14:42:02.508543) 167.793
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:02.710678(2022-01-04 14:42:02.709826) 169.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:02.910838(2022-01-04 14:42:02.909959) 172.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:03.111087(2022-01-04 14:42:03.110234) 173.793
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:03.311272(2022-01-04 14:42:03.310334) 175.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:03.512417(2022-01-04 14:42:03.511484) 178.293
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:03.713621(2022-01-04 14:42:03.712676) 179.757
TE:NDLT1542:MOT:MTR0101.RBV    2022-01-04 14:42:03.913854(2022-01-04 14:42:03.912849) 180
TE:NDLT1542:MOT:MTR0101        2022-01-04 14:42:03.913854(2022-01-04 14:42:03.913919) 180
TE:NDLT1542:MOT:MTR0101.DMOV   2022-01-04 14:42:03.913854(2022-01-04 14:42:03.914872) 1
TE:NDLT1542:ASTAXES_1:STSTATUS-BMOVING.RVAL 2022-01-04 14:42:03.944912(2022-01-04 14:42:03.945870) 0
```
looks like the DMOV is more closely related to MTR0101 value change, which looks like it only gets updated when the SP is reached, so could be the difference between that and bmoving changing + updating from the PLC. I think there's a minor difference in the poll rates of tcioc and the motor record scan loop (300ms and 200ms respectively) so that could be it. I can try with a faster tcioc rate like 150ms to see if that solves it

**Comment by rerpha** [comment]
yes it should, https://github.com/ISISComputingGroup/IBEX/issues/4669 is to sort that out 
for now not much of a problem as nowhere is using two of them, but will quickly become an issue if they do.

**Comment by rerpha** [comment]
Upping the scan rate for tcioc seems to have inadvertently fixed the MOVN problem as well, weird. Think that means #6839 may work once this is in?

**Comment by rerpha** [comment]
definitely. I am a bit stuck for ideas for it besides modding tcioc to shorten the names and just use the last part of the pv after the hyphen, like BENABLE, BENABLED etc.

**Comment by FreddieAkeroyd** [comment]
I guess we couldn't ask for shorter names in the axis structure :-D Will add some ideas to #4669
